### DISCUSSION TITLE
Ghidra fails to build on JDK 21

### DISCUSSION THREAD

**Discussion by agatti** [NationalSecurityAgency/ghidra#5968] [General]
I know Java 21 is not supported for the time being, this is just a heads up on potential breakages coming up in the future.  Not sure whether filing a bug for something that is explictly not supported is the right choice, so I'm posting here instead.

Anyway, with OpenJDK 21 on Linux/x64, `gradle buildGhidra` (using current Git HEAD - eaf6f0e6461642ff8bde6c7606559fcf416d83f4) fails with this:

```
> Task :Generic:compileJava
.../Ghidra/Framework/Generic/src/main/java/ghidra/generic/util/datastruct/TreeValueSortedMap.java:653: error: types List<Entry<K,V>> and Deque<Entry<K,V>> are incompatible;
        protected class ValueSortedTreeMapEntrySet extends AbstractSet<Entry<K, V>>
                  ^
  both define reversed(), but with unrelated return types
  where K,V are type-variables:
    K extends Object declared in class TreeValueSortedMap
    V extends Object declared in class TreeValueSortedMap
.../Ghidra/Framework/Generic/src/main/java/ghidra/generic/util/datastruct/TreeValueSortedMap.java:973: error: types List<K> and Deque<K> are incompatible;
        protected class ValueSortedTreeMapKeySet extends AbstractSet<K>
                  ^
  both define reversed(), but with unrelated return types
  where K is a type-variable:
    K extends Object declared in class TreeValueSortedMap
.../Ghidra/Framework/Generic/src/main/java/ghidra/generic/util/datastruct/TreeValueSortedMap.java:1211: error: types Deque<V> and List<V> are incompatible;
        protected class ValueSortedTreeMapValues extends AbstractCollection<V>
                  ^
  both define reversed(), but with unrelated return types
  where V is a type-variable:
    V extends Object declared in class TreeValueSortedMap
.../Ghidra/Framework/Generic/src/main/java/ghidra/generic/util/datastruct/ValueSortedMap.java:39: error: types Deque<Entry<K,V>> and List<Entry<K,V>> are incompatible;
        public interface ValueSortedMapEntryList<K, V>
               ^
  both define reversed(), but with unrelated return types
  where K,V are type-variables:
    K extends Object declared in interface ValueSortedMapEntryList
    V extends Object declared in interface ValueSortedMapEntryList
.../Ghidra/Framework/Generic/src/main/java/ghidra/generic/util/datastruct/ValueSortedMap.java:47: error: types Deque<K> and List<K> are incompatible;
        public interface ValueSortedMapKeyList<K> extends Set<K>, List<K>, Deque<K> {
               ^
  both define reversed(), but with unrelated return types
  where K is a type-variable:
    K extends Object declared in interface ValueSortedMapKeyList
.../Ghidra/Framework/Generic/src/main/java/ghidra/generic/util/datastruct/RestrictedValueSortedMap.java:253: error: types List<Entry<K,V>> and Deque<Entry<K,V>> are incompatible;
        public class RestrictedValueSortedMapEntryList implements ValueSortedMapEntryList<K, V> {
               ^
  both define reversed(), but with unrelated return types
  where K,V are type-variables:
    K extends Object declared in class RestrictedValueSortedMap
    V extends Object declared in class RestrictedValueSortedMap
.../Ghidra/Framework/Generic/src/main/java/ghidra/generic/util/datastruct/RestrictedValueSortedMap.java:548: error: types List<K> and Deque<K> are incompatible;
        public class RestrictedValueSortedMapKeyList implements ValueSortedMapKeyList<K> {
               ^
  both define reversed(), but with unrelated return types
  where K is a type-variable:
    K extends Object declared in class RestrictedValueSortedMap
```

**Comment by ryanmkurtz** [comment]
Thanks, we have an internal ticket to address it.

**Reply by ryanmkurtz** [reply]
See #6083.  Fixed by d92932a44d31fe355ae17e1e22e39ad584256b77.

**Comment by nuschpl** [comment]
Does this ChatGPT resolution helps in the issue:


> Here’s a breakdown of the issues:

> Method Conflicts: By extending both List and Deque, you are forcing ValueSortedMapEntryList<K, V> and ValueSortedMapKeyList<K> to adhere to the contracts of both interfaces. If there are methods that have the same name but different return types or parameters in List and Deque, it will lead to incompatibility issues.

> Design Consideration: Combining List and Deque into a single interface is generally not recommended because they represent different collection behaviors. A List is an ordered collection that allows precise control over where elements are inserted, while a Deque is a double-ended queue that allows elements to be added or removed from both ends.

> Resolving the Issue: To resolve this design issue, you might need to reconsider the structure of your ValueSortedMapEntryList<K, V> and ValueSortedMapKeyList<K> interfaces. It could be more effective to separate the functionalities of List and Deque or to find an alternative design that doesn't require extending both interfaces simultaneously.

> In summary, the source of your error is likely related to the complex inheritance and interface implementation in your ValueSortedMap design. To fix the error, you will need to revisit and possibly redesign how these interfaces interact and are extended in your code.
